src/QsCompiler/Transformations/LiftLambdas.cs

```diff
@@ -0,0 +1,257 @@
+                    : base(parent)
+                {
+                }
+
+                /// <inheritdoc/>
+                public override TypedExpression OnTypedExpression(TypedExpression ex)
+                {
+                    if (ex.Expression is ExpressionKind.Lambda lambda)
+                    {
+                        return this.HandleLambdas(ex, lambda.Item);
+                    }
+                    else
+                    {
+                        return base.OnTypedExpression(ex);
+                    }
+                }
+
+                private ParameterTuple MakeLambdaParams(ResolvedType expressionType, QsSymbol paramNames)
+                {
+                    var missingSymbolCount = 0;
+
+                    ParameterTuple MatchNameWithType(ResolvedType paramType, QsSymbol paramName)
+                    {
+                        if (paramName.Symbol.IsMissingSymbol)
+                        {
+                            var localVar = new LocalVariableDeclaration<QsLocalSymbol>(
+                                QsLocalSymbol.NewValidName($"__missingLambdaParam_{missingSymbolCount}__"),
+                                paramType,
+                                new InferredExpressionInformation(false, false),
+                                QsNullable<Position>.Null,
+                                paramName.Range.IsNull
+                                    ? DataTypes.Range.Zero
+                                    : paramName.Range.Item);
+                            missingSymbolCount++;
+                            return ParameterTuple.NewQsTupleItem(localVar);
+                        }
+                        else if (paramName.Symbol is QsSymbolKind<QsSymbol>.Symbol sym)
+                        {
+                            var localVar = new LocalVariableDeclaration<QsLocalSymbol>(
+                                QsLocalSymbol.NewValidName(sym.Item),
+                                paramType,
+                                new InferredExpressionInformation(false, false),
+                                QsNullable<Position>.Null,
+                                paramName.Range.IsNull
+                                    ? DataTypes.Range.Zero
+                                    : paramName.Range.Item);
+                            return ParameterTuple.NewQsTupleItem(localVar);
+                        }
+                        else if (paramName.Symbol is QsSymbolKind<QsSymbol>.SymbolTuple tup)
+                        {
+                            if (tup.Item.Length == 0 && paramType.Resolution.IsUnitType)
+                            {
+                                // Need an artificial Unit parameter here
+                                var localVar = new LocalVariableDeclaration<QsLocalSymbol>(
+                                    QsLocalSymbol.NewValidName("__lambdaUnitParam__"),
```

Comment:
Is this to handle `() -> ...` lambdas? The declarations `function F() : Unit {}` and `function F(x : Unit) : Unit {}` should be equivalent in Q#, is the tuple item needed?

Reply:
Consider the case:
```qsharp
let x = 3;
let lambda = () => x;
```
Here the parameter tuple needs to be mapped to a parameter `__lambdaUnitParam__` so that it can appear in the generated parameter list as `<lambda name>(x : Int, __lambdaUnitParam__ : Unit) : Int { return x; }` so that the original position of the lambda can be a partial application: `let lembda = <lambda name>(x, _);` Without this parameter the original expressions would be a completed call rather than the partial it needs to be.

Reply:
It _does_ also mean that `() => ...` gets mapped to `<lambda name>(__lambdaUnitParam__ : Unit)` and the call `<lambda name>(_)` when there is no closure, which is equivalent, as you point out, to the lambda if it were defined and called with no parameters, but this is a relatively harmless side effect, I think.